bundles/org.eclipse.swt/Eclipse SWT/gtk/org/eclipse/swt/graphics/Image.java

```diff
@@ -307,15 +309,28 @@ public Image(Device device, Image srcImage, int flag) {
 				byte[] line = new byte[stride];
 				for (int y=0; y<height; y++) {
 					C.memmove(line, data + (y * stride), stride);
-					for (int x=0, offset=0; x<width; x++, offset += 4) {
+					for (int x = 0, offset = 0; x < width; x++, offset += 4) {
 						int a = line[offset + oa] & 0xFF;
 						int r = line[offset + or] & 0xFF;
 						int g = line[offset + og] & 0xFF;
 						int b = line[offset + ob] & 0xFF;
-						line[offset + oa] = (byte) Math.round((double) a * 0.5);
-						line[offset + or] = (byte) Math.round((double) r * 0.5);
-						line[offset + og] = (byte) Math.round((double) g * 0.5);
-						line[offset + ob] = (byte) Math.round((double) b * 0.5);
+						// The alpha value is embedded into the RGB values as well, so extract it out
+						// of those values for transformation and reapply it afterwards
+						if (hasAlpha && a != 0) {
+							r = r * 255 / a;
+							g = g * 255 / a;
+							b = b * 255 / a;
+						}
+						RGBA result = DEFAULT_DISABLED_IMAGE_TRANSFORMER.adaptPixelValue(r, g, b, a);
+						if (hasAlpha) {
+							result.rgb.red = result.rgb.red * result.alpha / 255;
```

Comment:
Just a nit-pick: this and a few other places could be slightly more compact.
```suggestion
							result.rgb.red *= result.alpha / 255;
```

Reply:
Note that this is integer arithmetics and, if I am not mistaken, the proposal changes the execution order.
My understanding is that:
```
result.rgb.red = result.rgb.red * result.alpha / 255 === result.rgb.red = (result.rgb.red * result.alpha) / 255
```
And
```
result.rgb.red *= result.alpha / 255 === result.rgb.red = result.rgb.red * (result.alpha / 255)
```

The latter will always yield zero if alpha < 255.

@HannesWell Am I missing something?

Reply:
> Note that this is integer arithmetics and, if I am not mistaken, the proposal changes the execution order.

Indeed, I think you are right. Please disregard my comment.